### DISCUSSION TITLE
Large graphs question

### DISCUSSION THREAD

**Discussion by gmacon** [Byron/dua-cli#211] [Q&A] [answered]
In the README, you give a known limitation as

> In interactive mode, the maximum amount of files is limited to 2^32 - 1 (`u32::max_value() - 1`) entries.
>  * One node is used as to 'virtual' root
>  * The actual amount of nodes stored might be lower, as there might be more edges than nodes, which are also limited by a `u32` (I guess)
>  * The limitation is imposed by the underlying `petgraph` crate, which declares it as `unsafe` to use u64 for instance.
>  * It's possibly *UB* when that limit is reached, however, it was never observed either.

I was surprised to read that petgraph doesn't support 64-bit indices, so I went and looked it up. It doesn't support `u64`, but you can use `usize`, which isn't *quite* as good, since you only get 64-bit indices on 64-bit platforms, but it seems to me that that would be the common case. Is there some reason I don't see not to use `usize`?

**Comment by Byron** [comment, marked as answer]
Thanks for sharing your results, I wasn't aware and never tried `usize`.

> Is there some reason I don't see not to use `usize`?

I don't think there is except for increased memory consumption. Thus far I only ever saw file systems with 7m entries or so which consumes less than a gigabyte of memory. This means we could have ~613 times more of that until the `u32` becomes a problem, which probably also takes about 500GB of RAM. Thus I'd think this issue can be solved in a couple of decades when computers have caught up :) - it's easy enough.
